### PR TITLE
Feature/OAS

### PR DESCRIPTION
**Pull request by furkansenharputlu**
This PR is to implement some of OAS support for Tyk API definition. When it is finalized, any Tyk API will be able to be exported in the form of an OAS documentation.

### CODE REVIEW

File: apidef/oas/authentication.go [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+	}
+
+	if a.JWT != nil {
+		a.JWT.ExtractTo(api)
+	}
+}
+
+type Token struct {
+	Enabled                 bool `bson:"enabled" json:"enabled"` // required
+	AuthSources             `bson:",inline" json:",inline"`
```
Comment (vaske) [inline apidef/oas/authentication.go:64]: is this by intention to miss bson name for this one?

Comment (furkansenharputlu) [reply apidef/oas/authentication.go:64]: Yes

File: gateway/api.go [hunk 2]
```diff
@@ -871,6 +931,7 @@ func handleDeleteAPI(apiID string) (interface{}, int) {
 	}
 
 	os.Remove(defFilePath)
+	os.Remove(defOASFilePath)
```
Comment (vaske) [inline gateway/api.go:934]: I see report here 
```
Check failure on line 933 in gateway/api.go
```

**Review by vaske — CHANGES_REQUESTED** [review]
Great work so far, there are some check failures, could you please address them?

**Review by vaske — APPROVED** [review]

**Comment by buger** [comment]
I have a feeling that majority of this code should live in the gateway. 
We also want to have OSS Gateway run using OAS API definitions, how we can make it happen?

**Comment by furkansenharputlu** [comment]
@buger Yes, the conversion functions should live in the gateway. However, storing and returning proper style for user should be handled on both dashboard users and gateway users. For OSS, we already have such a logic where we store OAS format of an api definition in json file in this PR.

File: gateway/api.go [hunk 3]
```diff
@@ -825,8 +868,35 @@ func handleAddOrUpdateApi(apiID string, r *http.Request, fs afero.Fs) (interface
...
 	}
 
+	// api
+	err, errCode := writeToFile(fs, newDef, newDef.APIID)
+	if err != nil {
+		return apiError(err.Error()), errCode
+	}
+
+	// oas
+	err, errCode = writeToFile(fs, &oasDoc, newDef.APIID+"-oas")
```
Comment (furkansenharputlu) [inline gateway/api.go:878]: @buger like here

**Review by buger — APPROVED** [review]

**Comment by furkansenharputlu** [comment]
Manually cherry-picked all commits to `release-4`

**Pull request merged**